File: csharp/11.ContainerWithMostWater.cs
namespace Leetcode.ContainerWithMostWater;

public class Solution {
    public int MaxArea(int[] height) {
        int left = 0;
        int right = height.Length - 1;
        int maxArea = -1;

        while (right > left) {
            int maxL = height[left];
            int maxR = height[right];
            int area = CalculateArea(left, right);
            if (area > maxArea)
                maxArea = area;

            if (maxL > maxR) {
                // start looking for a new maxR
                while (right > left) {
                    right--;
                    if (height[right] > maxR) break;
                }
            } else {
                // start looking for a new maxL
                while (right > left) {
                    left++;
                    if (height[left] > maxL) break;
                }
            }
        }

        return maxArea;

        int CalculateArea(int p1, int p2) {
            return Math.Min(height[p1], height[p2]) * Math.Abs(p1 - p2);
        }
    }
}

File: golang/11.ContainerWithMostWater.go
package leetcode

func maxArea(height []int) int {
	left := 0
	right := len(height) - 1
	maxArea := -1

	for right > left {
		maxL := height[left]
		maxR := height[right]
		area := calcArea(height, left, right)
		if area > maxArea {
			maxArea = area
		}

		if maxL > maxR {
			for right > left {
				right--
				if height[right] > maxR {
					break
				}
			}
		} else {
			for right > left {
				left++
				if height[left] > maxL {
					break
				}
			}
		}
	}

	return maxArea
}

func calcArea(height []int, left int, right int) int {
	return min(height[left], height[right]) * abs(left-right)
}

func min(a, b int) int {
	if a >= b {
		return b
	}
	return a
}

func abs(v int) int {
	if v < 0 {
		return v * -1
	}
	return v
}
